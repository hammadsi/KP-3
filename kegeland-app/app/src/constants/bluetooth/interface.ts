import {DeviceStackParamList} from '~routes/interface';

export type BluetoothCharacteristic = string;
export type BluetoothService = {
  uuid: string;
  characteristics: BluetoothCharacteristic[];
};

export type BluetoothProfile = {
  navScreen: keyof DeviceStackParamList;
  services: string[];
  batteryService: {
    service: BluetoothService;
    fn: (bytes: number[]) => number;
  };
};

export type PeripheralNotification = {
  value: number[];
  peripheral: string;
  characteristic: string;
  service: string;
};
