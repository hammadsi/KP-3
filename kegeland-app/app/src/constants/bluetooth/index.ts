import FEMFIT_PROFILE from '~lib/femfit/bluetooth/profile';

import {BluetoothProfile} from './interface';

/**
 * How long to scan for bluetooth peripherals, in seconds.
 */
export const SCAN_TIME = 5;
/**
 * Whether or not to allows duplicate devices
 */
export const ALLOW_DUPLICATE_DEVICES = false;
/**
 * How often to read BLE characteristic notifications
 */
export const UPDATE_INTERVAL_MS = 250;

export const _profiles = {
  femfit: FEMFIT_PROFILE,
} as const;

export type DeviceType = keyof typeof _profiles;

/**
 * Record of supported bluetooth profiles
 */
export const BLE_PROFILES = _profiles as Record<DeviceType, BluetoothProfile>;

/**
 * Record of supported peripherals
 */
export const PERIPHERAL_MAP: Record<string, DeviceType> = {
  [FEMFIT_PROFILE.batteryService.service.uuid]: 'femfit',
};
