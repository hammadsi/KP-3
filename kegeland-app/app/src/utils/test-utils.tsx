/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
