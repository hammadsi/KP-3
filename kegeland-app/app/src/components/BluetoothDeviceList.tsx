import React from 'react';
import {StyleSheet, View} from 'react-native';
import {List, Text, useTheme} from 'react-native-paper';

import useAppSelector from '~hooks/useAppSelector';
import {BluetoothDevice} from '~state/ducks/bluetooth/bluetooth.interface';

import BluetoothDeviceItem from './BluetoothDeviceItem';
import Icon from './Icon';
import ListItemSkeleton from './ListItemSkeleton';

export type BluetoothDeviceListProps = {
  devices: BluetoothDevice[];
};

/**
 * Component for rendering a list of bluetooth devices.
 * @param props the props
 * @see {@link BluetoothDeviceListProps}
 */
const BluetoothDeviceList: React.FC<BluetoothDeviceListProps> = ({devices}) => {
  const {colors} = useTheme();
  const {isScanning} = useAppSelector((state) => state.bluetooth);
  const render = () => {
    if (!isScanning && devices.length === 0) {
      // Render a message if the list is empty
      return (
        <View style={styles.errorWrapper}>
          <Icon
            style={styles.errorIcon}
            color={colors.primary}
            icon="alert-circle-outline"
            size={32}
          />
          <Text style={styles.errorText}>No devices found!</Text>
        </View>
      );
    } else {
      return devices.length > 0
        ? devices.map((device) => (
            <BluetoothDeviceItem key={device.id} device={device} />
          ))
        : Array.from({length: 3}, (_v, i) => <ListItemSkeleton key={i} />);
    }
  };
  return <List.Section>{render()}</List.Section>;
};

const styles = StyleSheet.create({
  errorWrapper: {
    padding: 10,
    flexDirection: 'column',
    justifyContent: 'center',
  },
  errorIcon: {
    flex: 1,
  },
  errorText: {
    flex: 1,
    paddingTop: 8,
    textAlign: 'center',
    fontSize: 16,
  },
});

export default BluetoothDeviceList;
