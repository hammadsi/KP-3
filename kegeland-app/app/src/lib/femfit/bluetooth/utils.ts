import {clamp, mean, round} from 'lodash';

import {PRESSURE_SENSOR_MAX_VAL, PRESSURE_SENSOR_MIN_VAL} from './constants';

const range = PRESSURE_SENSOR_MAX_VAL - PRESSURE_SENSOR_MIN_VAL;

/**
 * Calculates the pressure increase relative to minimum value
 * @param value pressure data
 * @returns pressure percent
 */
export const pressurePercent = (value: number) =>
  (value - PRESSURE_SENSOR_MIN_VAL) / range;

/**
 * Converts byte data to a battery level
 * @param bytes byte data
 * @returns
 */
export const readBattery = (bytes: number[]) => {
  const voltage =
    // eslint-disable-next-line no-bitwise
    (((((bytes[4] & 255) << 8) | (bytes[5] & 255)) * 1.2) / 1024 / 220000) *
    1220000;

  return clamp(round(voltage / 6.1, 2), 0, 1);
};

/**
 * Converts bytes to a pressure level
 * @param byte1 first byte
 * @param byte2 second byte
 * @returns pressure level
 */
const convertPressure = (byte1: number, byte2: number) => {
  return clamp(
    // eslint-disable-next-line no-bitwise
    ((((byte1 & 255) << 8) | (byte2 & 255)) + 65536) * 0.00750063755,
    PRESSURE_SENSOR_MIN_VAL,
    PRESSURE_SENSOR_MAX_VAL,
  );
};

/**
 * Converts byte to temperature
 * @param byte temperature byte
 * @returns temperature
 */
const convertSensorTemperature = (byte: number) => {
  return byte * 0.16 + 10.0;
};

/**
 * Converts byte data from the femfit sensor to pressures and temperatures for each of the 8 sensors
 * @param data byte data
 * @returns list of pressures and list of temperatures
 */
export const readSensorBytes = (data: number[][]) => {
  const pressures: number[] = [];
  const temps: number[] = [];
  for (const bytes of data) {
    pressures.push(
      ...[
        round(
          mean([
            convertPressure(bytes[2], bytes[3]),
            convertPressure(bytes[5], bytes[6]),
            convertPressure(bytes[8], bytes[9]),
          ]),
          2,
        ),
        round(
          mean([
            convertPressure(bytes[11], bytes[12]),
            convertPressure(bytes[14], bytes[15]),
            convertPressure(bytes[17], bytes[18]),
          ]),
          2,
        ),
      ],
    );
    temps.push(
      ...[
        round(
          mean([
            convertSensorTemperature(bytes[4]),
            convertSensorTemperature(bytes[7]),
            convertSensorTemperature(bytes[10]),
          ]),
          2,
        ),
        round(
          mean([
            convertSensorTemperature(bytes[13]),
            convertSensorTemperature(bytes[16]),
            convertSensorTemperature(bytes[19]),
          ]),
          2,
        ),
      ],
    );
  }
  return {pressures, temperatures: temps};
};
