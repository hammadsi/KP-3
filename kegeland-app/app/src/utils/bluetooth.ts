import {chain, keys} from 'lodash';
import {startNotification, stopNotification} from 'react-native-ble-manager';

import {BLE_PROFILES, PERIPHERAL_MAP, DeviceType} from '~constants/bluetooth';
import {BluetoothService} from '~constants/bluetooth/interface';

/**
 * Retrieves a bluetooth profile for the specified device
 * @param key the device to retrieve
 * @returns the bluetooth profile for specified device
 */
export const getProfile = (key: DeviceType) => {
  return BLE_PROFILES[key];
};

/**
 * Reads the device battery using a custom function specified in the device's bluetooth profile.
 * @param key the device to use
 * @param bytes the bytes to read
 * @returns the battery state
 */
export const readPeripheralDeviceBattery = (key: DeviceType, bytes: number[]) =>
  getProfile(key).batteryService.fn(bytes);

/**
 * Retrieves all service ids supported
 * @returns all supported service ids
 */
export const getAllServiceIds = () => {
  return keys(PERIPHERAL_MAP);
};

/**
 * Retrieves the device screen for a selected device
 * @param key the device
 * @returns the specific screen for the specified device
 */
export const getDeviceScreen = (key: DeviceType) => getProfile(key).navScreen;

/**
 * Will retrieve a device type based on a list of service uuids
 * @param serviceUUIDs the service uuids to check
 * @returns the device type
 */
export const getPeripheralType = (serviceUUIDs: string[]) => {
  const key = chain(serviceUUIDs)
    .filter((uuid) => uuid in PERIPHERAL_MAP)
    .first()
    .value();
  return key ? PERIPHERAL_MAP[key] : undefined;
};

/**
 * Will add notify-listeners for all characteristics within a BLE service
 * @param peripheral the peripheral id
 * @param service the service to listen to
 */
export const addServiceListener = async (
  peripheral: string,
  service: BluetoothService,
) => {
  try {
    for (const char of service.characteristics) {
      await startNotification(peripheral, service.uuid, char);
    }
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
    throw new Error('An unknown error has occurred');
  }
};

/**
 * Will remove notify-listeners for all characteristics within a BLE service
 * @param peripheral the peripheral id
 * @param service the service to remove
 */
export const removeServiceListener = async (
  peripheral: string,
  service: BluetoothService,
) => {
  try {
    for (const char of service.characteristics) {
      await stopNotification(peripheral, service.uuid, char);
    }
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
    throw new Error('An unknown error has occurred');
  }
};
