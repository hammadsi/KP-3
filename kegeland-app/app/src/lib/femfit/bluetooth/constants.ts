import {BluetoothService} from '~constants/bluetooth/interface';
import {transformId} from '~utils/platform.utils';

export const PRESSURE_SENSOR_MIN_VAL = 760;
export const PRESSURE_SENSOR_MAX_VAL = 970;
export const ACTIVATION_THRESHOLD = 0.15;

/**
 * Transform UUIDs based on the current platform
 */
const HOUSEKEEPING_SERVICE_UUID = transformId(
  'b92a0001-4bf9-4870-8aa1-881b3a20ada4',
);
const HOUSEKEEPING_CHARACTERISTIC_UUID = transformId(
  'b92a0002-4bf9-4870-8aa1-881b3a20ada4',
);
const SENSOR_SERVICE_UUID = transformId('0d9e0001-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_1_UUID = transformId('0d9e0002-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_2_UUID = transformId('0d9e0003-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_3_UUID = transformId('0d9e0004-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_4_UUID = transformId('0d9e0005-c111-49cd-bba3-85c7471cb6fa');

/**
 * The femfit services
 */
export const SERVICES = {
  HOUSEKEEPING_SERVICE: HOUSEKEEPING_SERVICE_UUID,
  SENSOR_SERVICE: SENSOR_SERVICE_UUID,
};

/**
 * The housekeeping service for femfit
 */
export const HOUSEKEEPING_SERVICE: BluetoothService = {
  uuid: HOUSEKEEPING_SERVICE_UUID,
  characteristics: [HOUSEKEEPING_CHARACTERISTIC_UUID],
};

/**
 * The sensor service for femfit
 */
export const SENSOR_SERVICE: BluetoothService = {
  uuid: SENSOR_SERVICE_UUID,
  characteristics: [SENSOR_1_UUID, SENSOR_2_UUID, SENSOR_3_UUID, SENSOR_4_UUID],
};
