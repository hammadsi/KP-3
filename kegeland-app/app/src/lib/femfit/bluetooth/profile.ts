import {BluetoothProfile} from '~constants/bluetooth/interface';

import {HOUSEKEEPING_SERVICE, SENSOR_SERVICE} from './constants';
import {readBattery} from './utils';

/**
 * The bluetooth profile for femfit
 */
export default {
  navScreen: 'Femfit',
  services: [HOUSEKEEPING_SERVICE.uuid, SENSOR_SERVICE.uuid],
  batteryService: {
    service: HOUSEKEEPING_SERVICE,
    fn: readBattery,
  },
} as BluetoothProfile;
