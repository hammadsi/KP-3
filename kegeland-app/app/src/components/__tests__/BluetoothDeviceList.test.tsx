import React from 'react';

import {deviceMock} from '~state/ducks/__mocks__/bluetooth.mocks';
import {render} from '~utils/test-utils';

import BluetoothDeviceList, {
  BluetoothDeviceListProps,
} from '../BluetoothDeviceList';

describe('Test BluetoothDeviceList-component', () => {
  it('should render correctly', () => {
    const props: BluetoothDeviceListProps = {
      devices: [deviceMock],
    };
    const component = <BluetoothDeviceList {...props} />;
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
