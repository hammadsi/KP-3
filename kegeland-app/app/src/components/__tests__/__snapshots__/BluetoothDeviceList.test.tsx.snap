// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test BluetoothDeviceList-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "marginVertical": 8,
        },
        undefined,
      ]
    }
    theme={
      Object {
        "animation": Object {
          "scale": 1,
        },
        "colors": Object {
          "accent": "#03dac4",
          "backdrop": "rgba(0, 0, 0, 0.5)",
          "background": "#f6f6f6",
          "border": "rgb(216, 216, 216)",
          "card": "rgb(255, 255, 255)",
          "disabled": "rgba(0, 0, 0, 0.26)",
          "elevation": Object {
            "elevation1": "#fff",
          },
          "error": "#B00020",
          "muted": "rgba(0, 0, 0, 0.54)",
          "notification": "#f50057",
          "onSurface": "#000000",
          "placeholder": "rgba(0, 0, 0, 0.54)",
          "primary": "#00b0ff",
          "status": Object {
            "alert": "#ffd600",
            "error": "#dd2c00",
            "success": "#64dd17",
          },
          "surface": "#ffffff",
          "text": "#000000",
        },
        "dark": false,
        "fonts": Object {
          "light": Object {
            "fontFamily": "System",
            "fontWeight": "300",
          },
          "medium": Object {
            "fontFamily": "System",
            "fontWeight": "500",
          },
          "regular": Object {
            "fontFamily": "System",
            "fontWeight": "400",
          },
          "thin": Object {
            "fontFamily": "System",
            "fontWeight": "100",
          },
        },
        "roundness": 15,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "padding": 8,
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "marginRight": 8,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#00b0ff",
                "fontSize": 24,
              },
              Array [
                Object {
                  "alignSelf": "center",
                },
                undefined,
              ],
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰂯
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 16,
                      "lineHeight": 16,
                      "marginTop": 2,
                      "marginVertical": 0,
                    },
                  ],
                ],
              ]
            }
          >
            Test device
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontSize": 8,
                  },
                  Array [
                    Object {
                      "alignSelf": "center",
                    },
                    Array [
                      Object {
                        "marginRight": 4,
                      },
                      Object {
                        "color": "#ffd600",
                      },
                    ],
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰝥
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "fontSize": 12,
                    "lineHeight": 14,
                  },
                ]
              }
            >
              Available
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignSelf": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Object {
                  "marginRight": 4,
                },
              ]
            }
          >
            100
            %
          </Text>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#64dd17",
                  "fontSize": 20,
                },
                Array [
                  Object {
                    "alignSelf": "center",
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            󰥈
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "position": "relative",
          }
        }
      >
        <View
          collapsable={false}
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "alignSelf": "center",
                    },
                    undefined,
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰇙
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
